src/hotspot/cpu/x86/stubGenerator_x86_64_adler.cpp

```diff
@@ -132,28 +139,72 @@ address StubGenerator::generate_updateBytesAdler32() {
   __ jcc(Assembler::aboveEqual, SKIP_LOOP_1A);
 
   __ align32();
-  __ bind(SLOOP1A);
-  __ vbroadcastf128(ydata, Address(data, 0), Assembler::AVX_256bit);
-  __ addptr(data, CHUNKSIZE);
-  __ vpshufb(ydata0, ydata, yshuf0, Assembler::AVX_256bit);
-  __ vpaddd(ya, ya, ydata0, Assembler::AVX_256bit);
-  __ vpaddd(yb, yb, ya, Assembler::AVX_256bit);
-  __ vpshufb(ydata1, ydata, yshuf1, Assembler::AVX_256bit);
-  __ vpaddd(ya, ya, ydata1, Assembler::AVX_256bit);
-  __ vpaddd(yb, yb, ya, Assembler::AVX_256bit);
-  __ cmpptr(data, end);
-  __ jcc(Assembler::below, SLOOP1A);
+  if (VM_Version::supports_avx512vl()) {
+    __ cmpl(s, VM_Version::avx3_threshold());
+    __ jcc(Assembler::belowEqual, SPRELOOP1A_AVX2);
+    // AVX2 performs better for smaller inputs because of leaner post loop reduction sequence..
+    __ cmpl(s, 128);
+    __ jcc(Assembler::belowEqual, SPRELOOP1A_AVX2);
+    __ vpxor(yb, yb, yb, Assembler::AVX_512bit);
+
+    __ bind(SLOOP1A_AVX3);
+      __ evpmovzxbd(ydata, Address(data, 0), Assembler::AVX_512bit);
+      __ vpaddd(ya, ya, ydata, Assembler::AVX_512bit);
+      __ vpaddd(yb, yb, ya, Assembler::AVX_512bit);
+      __ addptr(data, CHUNKSIZE);
+      __ cmpptr(data, end);
```

Comment:
This still processes 16 bytes worth of data in one loop iteration as the AVX2 loop. Have you given thoughts on processing double the size with AVX3?

Reply:
If we double the size of inputs processed within the loop from 16 to 32 then subsequent ADLER_SCALE will also change from 0-31, given that we cannot accommodate more than 16 integer in a 512 bit vector hence it will not be possible to scale up the loop block size.

Reply:
Could be done something like below:
[adler32.patch](https://github.com/openjdk/jdk/files/10521481/adler32.patch)

Reply:
Yes, one level of unrolling will save loop jump back costs, I have incorporated this in the patch.